src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/ReadOnlyTensorSpan.cs

```diff
@@ -655,34 +655,70 @@ public ReadOnlyTensorSpan<T> Slice(params scoped ReadOnlySpan<NRange> ranges)
             if (ranges.Length != Lengths.Length)
                 throw new ArgumentOutOfRangeException(nameof(ranges), "Number of dimensions to slice does not equal the number of dimensions in the span");
 
+            ReadOnlyTensorSpan<T> toReturn;
             scoped Span<nint> lengths;
             scoped Span<nint> offsets;
+            nint[]? lengthsArray;
+            nint[]? offsetsArray;
             if (Rank > TensorShape.MaxInlineRank)
             {
-                lengths = stackalloc nint[Rank];
-                offsets = stackalloc nint[Rank];
+                lengthsArray = ArrayPool<nint>.Shared.Rent(Rank);
+                lengths = lengthsArray;
+                lengths = lengths.Slice(0, Rank);
+                lengthsArray = null;
+
+                offsetsArray = ArrayPool<nint>.Shared.Rent(Rank);
+                offsets = offsetsArray;
+                offsets = offsets.Slice(0, Rank);
+                offsetsArray = null;
             }
             else
             {
-                lengths = new nint[Rank];
-                offsets = new nint[Rank];
+                lengths = stackalloc nint[Rank];
+                offsets = stackalloc nint[Rank];
+
+                lengthsArray = null;
+                offsetsArray = null;
             }
 
             for (int i = 0; i < ranges.Length; i++)
             {
                 (offsets[i], lengths[i]) = ranges[i].GetOffsetAndLength(Lengths[i]);
             }
 
-            nint index = 0;
-            for (int i = 0; i < offsets.Length; i++)
+            // When we have an empty Tensor and someone wants to slice all of it, we should return an empty Tensor.
+            if (TensorSpanHelpers.CalculateTotalLength(Lengths) == 0)
             {
-                index += Strides[i] * (offsets[i]);
+                for (int i = 0; i < offsets.Length; i++)
+                {
+                    if (offsets[i] != 0 || lengths[i] != 0)
+                        ThrowHelper.ThrowIndexOutOfRangeException();
+                }
+                toReturn = new ReadOnlyTensorSpan<T>(ref _reference, lengths, _shape.Strides, _shape._memoryLength);
+            }
+            else
```

Comment:
The problem with doing that approach is that when the length is 0 and we try to slice it, this check:
`if ((nuint)offsets[i] >= (nuint)Lengths[i])`, will cause an error to be thrown. So we could modify that check to work with an extra condition, but I figured it would be better to separate it out so we only do the 0 check once.

Reply:
> So we could modify that check to work with an extra condition, but I figured it would be better to separate it out so we only do the 0 check once.

Doesn't doing the 0 check mean though looping through the whole thing and summing up the lengths anyway?

Reply:
Yeah. I have combined the checks now to avoid the dedicated path.